epifor/data/batch.py

```diff
@@ -0,0 +1,223 @@
+    BATCH_FILE_NAME = "batch.yaml"
+
+    config = jo.DictProperty()
+    comment = jo.StringProperty()
+    created = jo.DateTimeProperty(required=True)
+    sims = jo.ListProperty(SimInfo)
+    name = jo.StringProperty(required=True)
+    # Map {region_key: {region estimates etc}}
+    region_data = jo.DictProperty()
+
+    @classmethod
+    def new(cls, config, suffix=None):
+        "Custom constructor (to avoid conflicts with loading from yaml)."
+        now = datetime.datetime.now().astimezone()
+        return cls(
+            config=config,
+            comment=f"{getpass.getuser()}@{socket.gethostname()}",
+            created=now,
+            sims=[],
+            name=f"batch-{now.isoformat()}" + (f"-{suffix}" if suffix else ""),
+            region_data={},
+        )
+
+    def save(self):
+        "Save batch metadata, autonaming the file."
+        fname = self.get_batch_file_path()
+        log.info(f"Writing batch metadata to {fname}")
+        with open(fname, "wt") as f:
+            yaml.dump(self.to_json(), f)
+
+    @classmethod
+    def load(cls, path):
+        "Load batch metadata from path. Does not load the Simulations (see `load_sims`)"
+        log.info(f"Reading batch metadata from {path}")
+        with open(path, "rt") as f:
+            d = yaml.load(f)
+        return Batch(d)
+
+    def get_batch_file_path(self):
+        return self.get_out_dir() / self.BATCH_FILE_NAME
+
+    def get_data_sims_dir(self):
+        "GleamViz data dir Path, checks existence."
+        p = Path(self.config["gleamviz_dir"]).expanduser() / "data" / "sims"
+        assert p.exists() and p.is_dir()
+        return p
+
+    def get_out_dir(self, create=True):
+        "Batch output dir Path, creates by default."
+        p = Path(self.config["output_dir"]).expanduser() / self.name
+        if create:
+            p.mkdir(parents=True, exist_ok=True)
+        assert p.exists()
+        return p
+
+    def add_simulation_info(self, sim: Simulation, name, group, color=None, style=None):
+        "Add sim info after batch creation (before running simulations)"
+        if style is None:
+            style = dict(DEFAULT_LINE_STYLE)
+        if color is not None:
+            style["color"] = color
+        bs = SimInfo(
+            id=sim.definition.get_id(), name=name, line_style=style, group=group,
+        )
+        bs.sim = sim
+        self.sims.append(bs)
+
+    def load_sims(self, allow_unfinished=False):
+        "Load simulation all batch simulations (optionally failing if any uncomputed)"
+        sims_dir = self.get_data_sims_dir()
+        for bs in self.sims:
+            sdir = sims_dir / f"{bs.id}.gvh5"
+            bs.sim = Simulation.load_dir(sdir, skip_unfinished=False)
+            if not bs.sim.has_result() and not allow_unfinished:
+                die(f"Simulation {bs.sim.name!r} in {sdir} does not have result")
+        with_res = sum(int(bs.sim.has_result()) for bs in self.sims)
+        log.info(
```

Comment:
Yeah, good point about performance! I would keep it here, though:
* to use mostly the same formatting in most places (not fully consistent here, TBH - I first avoided f-strings to allow Py35)
* to avoid positional formatting parameters like a plague ;)